### PR TITLE
perf(ext/url): lazy cache for URL keys

### PR DESCRIPTION
**Pull request by littledivy**
This commits adds a key cache for URL parts. This has real performance
impact, like in Fresh -> https://github.com/denoland/fresh/pull/368

```
# This patch
url_parse:              n = 50000, dt = 0.043s, r = 1162791/s, t = 859ns/op
url_parse_get_path:     n = 50000, dt = 0.002s, r = 25000000/s, t = 40ns/op
```

```
# main
url_parse:              n = 50000, dt = 0.066s, r = 757576/s, t = 1320ns/op
url_parse_get_path:     n = 50000, dt = 0.002s, r = 25000000/s, t = 40ns/op
```

### CODE REVIEW

File: ext/url/00_url.js [hunk 1]
```diff
@@ -436,20 +435,21 @@
         prefix,
         context: "Argument 1",
       });
-      this[_url] = opUrlParse(value);
+      this[_rid] = opUrlParse(value);
+      this[_url] = new Array(11);
```
Comment (aapoalas) [inline ext/url/00_url.js:439]: This will create a holey array. I'm not sure how much worse that is in terms of access-by-index performance, but I would consider doing either a manual creation:
```js
this[_url] = [undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined];
```
or possibly having such an array in the module scope and making a copy of it here.

File: ext/url/lib.rs [hunk 2]
```diff
@@ -35,31 +39,24 @@ pub fn init() -> Extension {
...
 /// Parse `UrlParseArgs::href` with an optional `UrlParseArgs::base_href`, or an
-/// optional part to "set" after parsing. Return `UrlParts`.
+/// optional part to "set" after parsing.
 #[op]
 pub fn op_url_parse(
+  state: &mut OpState,
   href: String,
   base_href: Option<String>,
-) -> Result<UrlParts, AnyError> {
+) -> Result<u32, AnyError> {
```
Comment (aapoalas) [inline ext/url/lib.rs:59]: Should this be `ResourceId`?

File: ext/url/lib.rs [hunk 3]
```diff
@@ -84,60 +82,101 @@ pub enum UrlSetter {
...
+    UrlSetter::Origin => return Ok(quirks::origin(&url)),
+  };
+  Ok(part.to_string())
 }
 
 #[op]
 pub fn op_url_reparse(
-  href: String,
+  state: &mut OpState,
+  rid: u32,
```
Comment (aapoalas) [inline ext/url/lib.rs:121]: Should this be a `ResourceId` as well?

File: ext/url/00_url.js (hunk 1 above)
Comment (littledivy) [reply ext/url/00_url.js:445]: good point, just went with manual declaration to avoid primordials

File: runtime/js/40_testing.js [hunk 4]
```diff
@@ -397,7 +397,11 @@
         const preResource = pre[resource];
         const postResource = post[resource];
         if (preResource === postResource) continue;
-        if (ArrayPrototypeIncludes(resourceAllowList, postResource)) continue;
+        if (
+          ArrayPrototypeIncludes(resourceAllowList, preResource || postResource)
```
Comment (aapoalas) [inline runtime/js/40_testing.js:401]: Is it possible that both `preResource` and `postResource` are non-zero / undefined? If so, then this `.includes(X || Y)` will only check for `X` and will ignore `Y`.

Comment (littledivy) [reply runtime/js/40_testing.js:401]: Yes we want it to `Y` i.e `postResource` if `X` is truthy.

Comment (aapoalas) [reply runtime/js/40_testing.js:401]: Hmm, this was a bit hard to understand but I managed to explain it to myself well enough to believe that this is fine.

Comment (littledivy) [reply runtime/js/40_testing.js:401]: yikes...i read my comment again. I meant "we want it to _ignore_ Y"

Comment (aapoalas) [reply runtime/js/40_testing.js:401]: Yeah, I managed to gather your meaning. It was the code that was throwing me on a loop :)

File: cli/tests/unit/http_test.ts [hunk 5]
```diff
@@ -1025,7 +1025,7 @@ Deno.test(
       delete resourcesAfter[Number(rid)];
     }
 
-    assertEquals(Object.keys(resourcesAfter).length, 1);
+    assertEquals(Object.keys(resourcesAfter).length, 3);
```
Comment (bartlomieju) [inline cli/tests/unit/http_test.ts:1028]: Why 3 here?

File: runtime/js/40_testing.js [hunk 6]
```diff
@@ -363,6 +364,12 @@
     }
   }
 
+  // Some resources cannot be explicitly closed and depend on
+  // garbage collection.
+  const resourceAllowList = [
+    "urlResource",
+  ];
```
Comment (bartlomieju) [inline runtime/js/40_testing.js:371]: This is really special case... I'm worried this will open doors for more special cases... Maybe we could store it inside `OpState` instead of `ResourceTable`?

File: tools/wpt/expectation.json [hunk 7]
```diff
@@ -2165,32 +2165,26 @@
...
-      "URL: Setting <non-spec:/.//p>.hostname = ''",
       "URL: Setting <foo://somehost/some/path>.pathname = '' Non-special URLs can have their paths erased",
       "URL: Setting <foo:///some/path>.pathname = '' Non-special URLs with an empty host can have their paths erased",
       "URL: Setting <file://monkey/>.pathname = '\\\\' File URLs and (back)slashes",
       "URL: Setting <file:///unicorn>.pathname = '//\\/' File URLs and (back)slashes",
       "URL: Setting <file:///unicorn>.pathname = '//monkey/..//' File URLs and (back)slashes",
       "URL: Setting <non-spec:/>.pathname = '/.//p' Serialize /. in path",
       "URL: Setting <non-spec:/>.pathname = '/..//p'",
-      "URL: Setting <non-spec:/>.pathname = '//p'",
-      "URL: Setting <non-spec:/.//>.pathname = 'p' Drop /. from path"
```
Comment (bartlomieju) [inline tools/wpt/expectation.json:2193]: Why are these tests getting enabled? Which changes caused it? Could we do that in a separate PR?

**Comment by littledivy** [comment]
No longer needed. See #15663
